h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/remote/SteamExecutorStarter.java

```diff
@@ -0,0 +1,169 @@
+import hex.tree.xgboost.XGBoostModel;
+import hex.tree.xgboost.exec.RemoteXGBoostExecutor;
+import org.apache.log4j.Logger;
+import water.H2O;
+import water.Job;
+import water.fvec.Frame;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SteamExecutorStarter implements SteamMessenger {
+
+    private static final Logger LOG = Logger.getLogger(SteamExecutorStarter.class);
+
+    /**
+     * Initialized by Service lookup
+     */
+    private static SteamExecutorStarter instance;
+    
+    public static SteamExecutorStarter getInstance() {
+        return instance;
+    }
+
+    private static class ClusterInfo {
+        final String uri;
+        final String userName;
+        final String password;
+
+        private ClusterInfo(String uri, String userName, String password) {
+            this.uri = uri;
+            this.userName = userName;
+            this.password = password;
+        }
+    }
+    
+    private final Object sendingLock = new Object[0];
+    private final Object clusterLock = new Object[0];
+
+    private SteamMessageSender sender;
+    private ClusterInfo cluster;
+
+    private final Map<String, Map<String, String>> receivedMessages = new HashMap<>();
+    
+    public SteamExecutorStarter() {
+        instance = this;
+    }
+
+    public RemoteXGBoostExecutor getRemoteExecutor(XGBoostModel model, Frame train, Job<XGBoostModel> job) throws IOException {
+        synchronized (clusterLock) {
+            if (cluster == null) {
+                LOG.info("Starting external cluster for model " + model._key + ".");
+                startCluster(job);
+            } else {
+                LOG.info("External cluster available, starting model " + model._key + " now.");
+            }
+            return makeExecutor(model, train);
+        }
+    }
+
+    private void startCluster(Job<XGBoostModel> job) throws IOException {
+        Map<String, String> startRequest = makeStartRequest();
+        sendMessage(startRequest);
+        Map<String, String> response = waitForResponse(startRequest, job);
+        if (response != null && response.get("status").equals("started")) {
+            String remoteUri = response.get("uri");
+            String userName = response.get("user");
+            String password = response.get("password");
+            cluster = new ClusterInfo(remoteUri, userName, password);
+            LOG.info("External cluster started at " + remoteUri + ".");
+        } else {
+            throw new IllegalStateException("Failed to start external cluster: " + response.get("reason"));
+        }
+    }
+
+    private RemoteXGBoostExecutor makeExecutor(XGBoostModel model, Frame train) {
+        return new RemoteXGBoostExecutor(model, train, cluster.uri, cluster.userName, cluster.password);
+    }
+    
+    private Map<String, String> waitForResponse(Map<String, String> startRequest, Job<XGBoostModel> job) {
```

Comment:
is it possible that waitForResponse() will be called before queueResponseMessage?

Reply:
yes, in fact that always happens, that's why I have the while loop waiting for queueResponseMessage to put the response into received messages

Reply:
I was thinking whether there could be possible deadlock because both methods have synchronized block on the same object but it was only my confusion how wait and notify works